osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs

```diff
@@ -142,7 +150,10 @@ private void onClick()
             else
             {
                 if (Room?.Host?.Equals(localUser) == true)
+                {
+                    gameplayStartTracker.BeginOperation();
```

Comment:
This is impossible to follow. I can't find the end call.

I was hoping for an interface using `IDisposable` to make following the lifetime of the operations feasible:

```csharp
    /// <summary>
    /// Utility class to track ongoing online operations' progress.
    /// Can be used to disable interactivity while waiting for a response from online sources.
    /// </summary>
    public class OngoingOperationTracker
    {
        /// <summary>
        /// Whether there is an online operation in progress.
        /// </summary>
        public IBindable<bool> InProgress => inProgress;

        private readonly Bindable<bool> inProgress = new BindableBool();

        private LeasedBindable<bool> leasedInProgress;

        /// <summary>
        /// Begins tracking a new online operation.
        /// </summary>
        /// <exception cref="InvalidOperationException">An operation has already been started.</exception>
        /// <returns>An <see cref="IDisposable"/> which should be disposed when the operation and completed.</returns>
        public IDisposable BeginOperation()
        {
            if (leasedInProgress != null)
                throw new InvalidOperationException("Cannot begin operation while another is in progress.");

            leasedInProgress = inProgress.BeginLease(true);
            leasedInProgress.Value = true;

            return new InvokeOnDisposal(endOperation);
        }

        /// <summary>
        /// Ends tracking an online operation.
        /// Does nothing if an operation has not been begun yet.
        /// </summary>
        private void endOperation()
        {
            leasedInProgress?.Return();
            leasedInProgress = null;
        }
    }
```

Reply:
You've picked the worst usage by far, which is fair, but I'm not sure how to use `IDisposable` there. That disposable, once taken by the ready button, would have to find its way to the parent screen so that it can release the disable, at which point the original problem of having to push it through however many levels of the hierarchy returns (and requires not accidentally dropping the disposable along the way).

Reply:
I'd hope the full lease logic would be at the parent in the first place. Is that not possible?

Reply:
Okay, so the ready button logic is *really* not how we want to do things. It shouldn't be interacting with the API directly - this code should be moved to the point of construction and the action should be passed into the button IMO.

The Ready button itself should only be responsible for its display logic (it's fine for it to bind to the client for this purpose).

Reply:
Yeah alright, I'll see what I can do. Let me get my bearings.